src/ProductManifold.jl

```diff
@@ -0,0 +1,416 @@
+
+function proj_product(x::ProductTVector, i::Integer)
+    return x.parts[i]
+end
+
+"""
+    ProductCoTVector(parts)
+
+A more general but slower representation of cotangent vectors on a product
+manifold.
+"""
+struct ProductCoTVector{TM<:Tuple} <: CoTVector
+    parts::TM
+end
+
+ProductCoTVector(covectors...) = ProductCoTVector{typeof(covectors)}(covectors)
+eltype(x::ProductCoTVector) = eltype(Tuple{map(eltype, x.parts)...})
+similar(x::ProductCoTVector) = ProductCoTVector(map(similar, x.parts)...)
+similar(x::ProductCoTVector, ::Type{T}) where T = ProductCoTVector(map(t -> similar(t, T), x.parts)...)
+
+function proj_product(x::ProductCoTVector, i::Integer)
+    return x.parts[i]
+end
+
+function isapprox(M::ProductManifold, x, y; kwargs...)
+    return all(t -> isapprox(t...; kwargs...), ziptuples(M.manifolds, x.parts, y.parts))
+end
+
+function isapprox(M::ProductManifold, x, v, w; kwargs...)
+    return all(t -> isapprox(t...; kwargs...), ziptuples(M.manifolds, x.parts, v.parts, w.parts))
+end
+
+function representation_size(M::ProductManifold, ::Type{T}) where {T}
+    return (mapreduce(m -> sum(representation_size(m, T)), +, M.manifolds),)
+end
+
+manifold_dimension(M::ProductManifold) = sum(map(m -> manifold_dimension(m), M.manifolds))
+
+struct ProductMetric <: Metric end
+
+function det_local_metric(M::MetricManifold{ProductManifold, ProductMetric}, x::ProductArray)
+    dets = map(det_local_metric, M.manifolds, x.parts)
+    return prod(dets)
+end
+
+function inner(M::ProductManifold, x, v, w)
+    subproducts = map(inner, M.manifolds, x.parts, v.parts, w.parts)
+    return sum(subproducts)
+end
+
+function exp!(M::ProductManifold, y, x, v)
+    map(exp!, M.manifolds, y.parts, x.parts, v.parts)
+    return y
+end
+
+function exp(M::ProductManifold, x::ProductMPoint, v::ProductTVector)
+    return ProductMPoint(map(exp, M.manifolds, x.parts, v.parts)...)
+end
+
+function log!(M::ProductManifold, v, x, y)
+    map(log!, M.manifolds, v.parts, x.parts, y.parts)
+    return v
+end
+
+function log(M::ProductManifold, x::ProductMPoint, y::ProductMPoint)
+    return ProductTVector(map(log, M.manifolds, x.parts, y.parts)...)
+end
+
+function distance(M::ProductManifold, x, y)
+    return sqrt(sum(map(distance, M.manifolds, x.parts, y.parts).^2))
+end
+
+function injectivity_radius(M::ProductManifold, x)
+    return min(map(injectivity_radius, M.manifolds, x.parts)...)
+end
+
+function injectivity_radius(M::ProductManifold)
+    return min(map(injectivity_radius, M.manifolds)...)
+end
+
+"""
+    ProductRetraction(retractions::AbstractRetractionMethod...)
+
+Product retraction of `retractions`. Works on [`ProductManifold`](@ref).
+"""
+struct ProductRetraction{TR<:Tuple} <: AbstractRetractionMethod
+    retractions::TR
+end
+
+ProductRetraction(retractions::AbstractRetractionMethod...) = ProductRetraction{typeof(retractions)}(retractions)
+
+function retract!(M::ProductManifold, y, x, v, method::ProductRetraction)
```

Comment:
Would be good to have a version that takes a usual `AbstractRetractionMethod` and applies it to each submanifold. I'm thinking this would be convenient for e.g. exponential retractions.

Reply:
If you need exponential retractions for each manifold then you can just use `exp`, right? I think it's hard to imagine a good use case for this.

Reply:
It's quite easy: Any exponential map is a retraction. But retractions are more general (they only approximate exp). They can, however, be much more easy to be computed (i.e. (`x+y/||x+y||` is a retraction on the sphere). For me they are (btw) just called retractions (not neccessarily exponential).
However, I ghink the `ProductRetracion` should be fine, since one can build such a product retraction from any set (vector) of single retraction ones?

Reply:
Yes, `ProductRetraction` is about specifying retractions for each manifold, then they used respectively.